### ISSUE TITLE
HoloMap indexing: snapping to nearest key value

### ISSUE THREAD

**Issue by drs251** [ioam/holoviews#2107]
Currently, one has to specify the precise key value in order to retrieve an element from a HoloMap, i.e. if HoloMap ```hmap``` has a key value of 6.0041544914297296, then one must type ```hmap[6.0041544914297296]```, while ```hmap[6]``` will produce an error.

I propose that the indexing of HoloMaps should snap to the nearest available key value, similar to the behavior of most other holoviews objects.

**Comment by jlstevens** [comment]
I agree consistent semantics are important but we would need to think carefully about the implications, especially with respect to backwards compatibility (it *should* be entirely backwards compatible, I think!). For example, if this were implemented ``__getitem__`` wouldn't emit ``KeyError``s for missing keys...

**Comment by drs251** [comment]
In my opinion, the added convenience and consistency would merit breaking backward compatibility a little bit (especially if the change happens for version 2.0). Or is there a lot of code which depends on ```KeyError```s being thrown?

**Comment by jlstevens** [comment]
> Or is there a lot of code which depends on ``KeyError``s being thrown?

Not that I know of! I think I do support this feature request and it should be relatively easy to do, using our existing utilities. It might be a little trickier for ``DynamicMap`` to snap to the values specified in the dimension objects (if any).

I *think* this can be viewed more as an enhancement than a backwards compatibility breaking change. That said, I would want to merge any PR adding this earlier to make sure to test the new behavior thoroughly before the next release.
